[2단계 - 출석 다시 구현하기] 벨로(방예혁) 미션 제출합니다.

YehyeokBang opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

안녕하세요 엘리!

모든 코드를 지우고 다시 시작할 때, 이전의 구조가 자꾸 맴돌아서
최대한 잊고 의식적으로 TDD 사이클을 경험하려고 해봤습니다!

이번 2단계 리뷰도 잘 부탁드립니다. 🙇🏻🙇🏻

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
  -  <img width="250" alt="image" src="https://github.com/user-attachments/assets/78c7a015-50ee-4be9-8606-31cc77744ec0" />
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [벨로의 프롤로그](https://prolog.techcourse.co.kr/studylogs/3908)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 3.5
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

최대한 엔티티를 작게 유지하려고 했으나, 개인적으로 기대한 만큼은 아닌 것 같았습니다.
그래도 이전 피드백을 바탕으로 개선하기 위해 노력했습니다. 감사합니다 😄 
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

### TDD 느낀점

> 요구 사항을 읽고 필요한 객체를 먼저 구상한 뒤, 작은 단위부터 고민하며 점진적으로 개선하는 방식으로 접근했습니다.

이번 도메인은 비교적 익숙했기 때문에 TDD를 적용하는 과정이 예상보다 수월했습니다. 
그러나 익숙하지 않거나 복잡한 도메인에서는, 
불필요한 기능(사실 나중에 필요 없었던 기능 등)까지 테스트를 작성하는 데,
많은 시간이 소모될 가능성이 있다고 생각합니다.

그럼에도 불구하고 TDD 사이클을 따라가면서 
객체의 역할과 책임을 명확하게 정의할 수 있다는 점이 특히 유용하게 느껴졌습니다. 👍 

### TDD와 커밋 단위에 대한 고민

> 이번 미션의 핵심 목표가 TDD였던 만큼, 의식적으로 사이클을 유지하려고 노력했습니다.

저는 기존에 **하나의 변경점 + 실행 가능한 단위(또는 에러가 없는 상태)** 를 커밋 단위로 생각했습니다.
그런데 TDD를 적용하면서 `Red, Green, Refactor 단계마다 커밋을 해야 할까?` 라는 고민이 생겼습니다.

예를 들어,

- `red: 닉네임은 중복되면 안 된다.`
- `green: 닉네임 중복 검증 구현`
- `refactor: 닉네임 중복 검증 기능 리팩토링`

이런 식으로 TDD의 단계를 커밋으로 남기는 게 좋을지 고민했지만,
`커밋 기록을 보는 사람 입장에서 TDD 사이클이 드러나는 것이 정말 중요할까?` 라는 의문이 들었습니다.
결국, 기존에 생각하던 **의미 있는 변경 단위**로 커밋을 진행했습니다.

만약 제가 커밋 단위 규칙을 정해야 한다면, (저도 지키려고 노력 중인) 아래 두 가지를 중요하게 생각할 것 같아요.

- `각 커밋이 논리적으로 완결된 변경 사항을 담고 있는가?`
- `커밋 메시지만 보고도 변경 내용을 쉽게 파악할 수 있는가?`

> 글을 작성하다가 든 생각인데, TDD 사이클을 연습하는 과정에서는 단계별로 커밋하고, 이후 스쿼시로 의미 있는 단위로 정리하는 방법도 좋은 전략이 될 수 있을 것 같네요. 😄 

협업을 위해 팀의 규칙을 지키는 것이 가장 중요하겠지만,
엘리가 커밋 단위 규칙을 세워야 한다면, 어떤 것을 중요하게 생각할 것 같으신가요?
또는 엘리가 느낀 커밋 단위에 대한 팁이 궁금해요! 😄 

### 출력 요구 사항과 메서드의 명확성

> 메서드의 역할과 변경 가능성에 대한 고민

처음에는 `edit()` 메서드가 단순히 수정 기능만 수행해야 한다고 생각하여 반환 타입을 `void`로 두었습니다.
이는 `add()`, `edit()` 같은 메서드가 반환값을 가지면,
`어떤 값을 반환하는지`, `반환된 값이 변경되면 어디까지 영향을 미치는지` 등의
관리 포인트가 증가한다고 생각하기 때문입니다.

그러나, 출력 요구 사항을 고려하면 수정 후 `많은 정보를 출력해야 하는 상황`입니다.
`12월 03일 화요일 10:07 (지각) -> 09:58 (출석) 수정 완료!`
- 수정 날짜(및 요일), 수정 전 시각, 수정 전 출석 유형, 수정 후 시각, 수정 후 출석 유형

그래서 edit() 메서드를 void로 두었을 때 컨트롤러에 아래와 같은 형태로 구현되었습니다.
```java
// 변경 전 상태 조회
AttendanceLog beforeLog = 닉네임과_날짜로_존재하는_로그_가져오기();
AttendanceType beforeType = 출석_유형_구하기(beforeLog);

// 수정 수행
출석부.edit(바꿀_출석_로그);

// 변경 후 상태 조회
AttendanceLog afterLog = 닉네임과_날짜로_존재하는_로그_가져오기();
AttendanceType afterType = 출석_유형_구하기(afterLog);

// 출력
print(beforeLog, beforeType, afterLog, afterType);
```

- 컨트롤러가 출석 유형을 직접 구해야 했어요. (흐름 관리 이상의 출석 유형 판단 로직까지 컨트롤러가 담당)
- 수정했을 때 이전->이후 형태로 어떻게 바뀌었는지를 반횐하는 것은 꽤 자연스럽지 않을까?

라는 마음으로 Dto에 담아 반환하도록 했습니다.

```java
EditResult editResult = 출석부.edit();
print(editResult);
```

장점
✅ 컨트롤러가 출석 유형을 직접 판단하지 않아도 된다.
✅ 코드가 더 깔끔해지고, 변경 전, 후 정보를 자연스럽게 활용할 수 있다.
✅ 수정하면 이전 상태를 반환하는 것이 개인적으로 어느 정도 자연스러우며, 납득할 수 있는 방식이다.

단점
❌ `edit()` 메서드가 단순히 수정만 하는 것이 아니라 출력용 데이터를 생성하는 역할까지 포함하게 된다.
❌ 출력 형식 변경에 취약하다. 예를 들어, "수정되었습니다!"만 출력하도록 변경된다면 도메인의 edit() 메서드 자체를 수정해야 한다.

현재 TDD에 집중하며 마감 기한을 고려한 결과,
`깔끔한 코드 + 납득할 만한 반환값`을 기준으로 DTO를 사용했습니다.

그러나, `도메인 로직과 출력 요구 사항의 결합도 증가`한다는 점은 여전히 고민이 됩니다.

> 지금까지는 현재 선택이 출력 요구 사항이 유지되는 한 가장 실용적인 방법이라고 생각했습니다.

제 생각 흐름에 대한 엘리의 조언이 필요할 것 같아요. 😢

YehyeokBang committed:
chore: step2 준비

- 모든 코드를 초기화합니다.

YehyeokBang committed:
docs: 구현할 기능 목록 작성

YehyeokBang committed:
feat: 출석 기록 모델 구현

feat: 출석 로그 닉네임 검증 구현

- 출석 로그는 닉네임과 출석한 날짜, 시간을 가집니다.

feat: 닉네임 객체로 분리

- 닉네임이 스스로 검증을 책임질 수 있도록 분리합니다.

feat: 결석 간주 기준 추가

- 출석 기록에 시간 기록이 없는 경우 결석으로 간주합니다.

YehyeokBang committed:
feat: 닉네임 및 출석 로그 equals 구현

- 닉네임 값과 출석 날짜가 같은 경우 동일한 출석 로그로 간주합니다.

YehyeokBang committed:
feat: 출석 로그 목록 모델 구현

- 출석 로그를 관리하는 역할을 수행합니다.

YehyeokBang committed:
feat: 출석 유형 판단 기능 구현

- 출석 유형은 열거형으로 관리합니다.

YehyeokBang committed:
feat: 교육 일정 모델 구현

- 요일로 교육 시작 시간을 조회합니다.

YehyeokBang committed:
feat: 주말 및 법정 공휴일에 출석 불가 처리 구현

- 캠퍼스 운영 시간이 아닌 경우에도 출석할 수 없습니다.

YehyeokBang committed:
refactor: 법정 공휴일 관련 기능 분리

- 다양한 법정 공휴일을 관리합니다.

YehyeokBang committed:
refactor: 메서드명 수정

- 기록된 시간이 없으면 결석으로 간주하지만,
  기록된 시간이 있어도 결석이 될 수 있기 때문에 메서드명을 수정합니다.
- 해당 메서드 테스트 코드를 추가합니다.

YehyeokBang committed:
feat: 출석 로그 파일 모델 구현

YehyeokBang committed:
feat: 명령어 모델 구현

- 각 기능을 대표하는 명령어를 추가합니다.

YehyeokBang committed:
feat: 명령어 입력 기능 구현

- 처음 기준 날짜로 환영 메시지를 출력합니다.

YehyeokBang committed:
feat: 출석 확인 기능 구현

- 닉네임과 등교 시간을 입력 받아 출석을 기록합니다.

YehyeokBang committed:
feat: 출석 수정 기능 구현

- 없는 날을 수정하려는 경우 요청한 날짜와 시간으로 새로운 기록을 추가합니다.

YehyeokBang committed:
feat: 크루별 출석 기록 확인 기능 구현

YehyeokBang committed:
feat: 제적 위험자 확인 기능 임시 구현

- 임시 구현입니다.
- 정렬 요구사항을 구현할 예정입니다.

YehyeokBang committed:
feat: 프로그램 종료 기능 구현

- 사용자가 종료할 때까지 종료되지 않습니다.

YehyeokBang committed:
feat: 명단 및 검증 로직 구현

- 존재하지 않는 크루의 닉네임을 사용할 경우 예외가 발생한다.

YehyeokBang committed:
feat: 예외 메시지 출력 기능 구현

YehyeokBang committed:
refactor: 컨트롤러 비즈니스 로직 분리

- 크루 명단과 출석 기록을 가지는 출석부를 구현하여 비즈니스 로직을 포장합니다.
- 기존 컨트롤러에서 출석 유형을 판단하는 등 비즈니스 로직을 분리합니다.

YehyeokBang committed:
docs: 실제 실행 예시로 변경

YehyeokBang committed:
refactor: 메서드 분리

- 들여쓰기 요구사항을 위해 수정합니다.

YehyeokBang committed:
refactor: 스트림에서 for 문으로 변경

- 스트림의 forEach 문이 동시에 상태를 변경할 때 발생할 수 있는 문제를 방지하기 위해 for 문으로 변경합니다.

YehyeokBang committed:
refactor: 정규식 패턴 객체 상수화

- 정규식이 반복 호출될 때 새로운 Pattern 객체를 생성하지 않도록 변경합니다.

YehyeokBang committed:
refactor: 닉네임 길이 조건 상수화

yebink commented on src/main/java/attendance/AttendanceApplication.java:
```diff
@@ -0,0 +1,17 @@
+    }
+}
+
+// TODO: 예외 메시지 처리, 예외 상황 검토, 요구사항 검토, 효율적인 자료 구조 고민
```
이건 완료된걸까요 ? 😅

YehyeokBang replied:
네 맞습니다.
TODO 주석으로 작성한 만큼 제출 전에 쉽게 찾아볼 수 있었는데,
익숙하지 않아서 적어두고 놓친 것 같네요. 😢
수정하겠습니다.

꼼꼼한 리뷰 감사합니다. 👍

yebink commented on src/main/java/attendance/dto/AttendResult.java:
```diff
@@ -0,0 +1,10 @@
+import attendance.model.AttendanceType;
+import java.time.LocalDateTime;
+
+public record AttendResult(
```
dto 클래스는 클래스명에 dto임을 표현해주세요~
```suggestion
public record AttendResultDto(
```

YehyeokBang replied:
더 명확한 방법인 것 같습니다! 감사합니다.

yebink commented on src/main/java/attendance/model/NicknameRoster.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Collection;
+import java.util.Set;
+
+public class NicknameRoster {
```
`NicknameRoaster` 라는 네이밍만으로는 어떤 클래스인지 추측이 어려운데, 
네이밍에 이유가 있으신가요~?

YehyeokBang replied:
CSV 파일에 존재하지 않는 닉네임은 기능을 사용할 수 없기 때문에,
닉네임이 등록되어 있는지 확인할 수 있는 (크루) `명단`을 표현하고 싶었습니다. 
(닉네임이 사전에 등록된 목록이라는 점을 강조하고 싶었습니다.)

다시 찾아보니 `Roster`가 교대나 서비스를 맡는 이들의 순번이나
교대 목록을 보여주는 리스트, 선수 명단 등의 의미를 가진만큼
객체가 가지는 역할을 애매하게 해석될 여지가 있는 것 같아요.

`NicknameRegistry`, `NicknameDirectory` 같은 이름을 사용하면,
닉네임이 등록된 목록이라는 점이 더 명확하게 전달될 수도 있을 것 같아요. 👍

yebink commented:
## TDD와 커밋 단위에 대한 고민
의미 있는 고민이네요 👍 
개인적으로 커밋 메시지에 TDD 사이클이 드러날 필요는 없다고 생각합니다. TDD 는 결국 단단한 설계와 코드 품질을 높이고자 선택하는 방식 중 하나이지, 누군가에게 TDD 로 작성되었음 을 보여주는 행위가 중요하다고 생각하지는 않습니다!

각 커밋이 논리적으로 완결된 변경 사항을 담고 있는가?
커밋 메시지만 보고도 변경 내용을 쉽게 파악할 수 있는가?
글을 작성하다가 든 생각인데, TDD 사이클을 연습하는 과정에서는 단계별로 커밋하고, 이후 스쿼시로 의미 있는 단위로 정리하는 방법도 좋은 전략이 될 수 있을 것 같네요. 😄

협업을 위해 팀의 규칙을 지키는 것이 가장 중요하겠지만,
엘리가 커밋 단위 규칙을 세워야 한다면, 어떤 것을 중요하게 생각할 것 같으신가요?
또는 엘리가 느낀 커밋 단위에 대한 팁이 궁금해요! 😄

yebink commented on src/main/java/attendance/model/AttendanceLogs.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    public void edit(AttendanceLog updateAttendanceLog) {
+        logs.remove(updateAttendanceLog);
+        logs.add(updateAttendanceLog);
```
이 로직만 보면 동일한 객체를 remove / add 하는 것 같은데, 실제로는 AttendanceLog 에 equals / hashCode 가 재정의되어 있기 때문에 닉네임 / 날짜 만 같은 기존 객체가 제거되고, 시간이 수정된 새로운 객체가 추가되는 방식이군요!

equals / hashCode 가 특정 요구사항에 핏하게 짜여진 것 같은데, 다른 곳에서 추가로 시간까지 비교해야 하는 상황이 오면 어떻게 될까요?

YehyeokBang replied:
> AttendanceLog 에 equals / hashCode 가 재정의되어 있기 때문에 닉네임 / 날짜 만 같은 기존 객체가 제거되고, 시간이 수정된 새로운 객체가 추가되는 방식이군요!

맞습니다! 👍 

> equals / hashCode 가 특정 요구사항에 핏하게 짜여진 것 같은데, ...

같은 날에는 2개의 출석 기록을 생성할 수 없기 때문에,
누가 이것을 책임져야 하는가?에 대한 고민이 있었습니다.

요구 사항에서는 최소한의 구조를 유지하라는 방향이 있어서,
우선 출석 기록(AttendanceLog) 자체가 중복 여부를 책임지도록 구현했습니다.

> 다른 곳에서 추가로 시간까지 비교해야 하는 상황이 오면 어떻게 될까요?

출석 로그 객체에서 시간까지 비교해야 하는 상황이 떠오르지 않는 것 같아요.
혹시 예를 들어주실 수 있나요? 😢 
다만, 핏한 equals는 문제가 될 여지가 있는 것 같습니다. 👍 

(우선 시간까지 비교해야 하는 상황이 예시로 떠오르지 않아서, 핏한 equals가 문제가 될 수 있는 상황을 고민해봤습니다!)

말씀해주신 대로 현재 AttendanceLogs가 AttendanceLog의 equals와 hashCode에 강하게 의존하고 있어, 
만약 조퇴나 퇴실(같은 날이지만, 추가가 필요한 상황) 같은 기능이 추가된다면, 
단순히 equals를 수정하는 것만으로 해결되지 않을 가능성이 크겠네요. 이 경우, 더 넓은 범위의 코드 수정이 필요할 것 같습니다.

애초에 `같은 날에는 2개의 출석 기록을 생성할 수 없다`는 정책을 개별 로그가 아닌,
로그 목록을 관리하는 쪽에서 검증하는 게 더 유연한 설계일 수도 있겠다는 생각이 드네요.

YehyeokBang replied:
### 리뷰 이후 생각 흐름 정리

AttendanceLog에서 equals와 hashCode를
nickname과 attendanceDate로 비교하게 구현했습니다. (핏하게)
이렇게 하면 같은 사람의 같은 날 출석 기록은 하나만 존재하도록 컬렉션을 활용하여 쉽게 중복을 방지할 수 있습니다.

초기에 해당 구조로 구현한 이유가 `출석 기록이 하루에 하나만 가능하다`는 도메인 규칙을
객체 수준에서 스스로 보장하게 하려는 의도였습니다.

### `출석 기록의 개별 객체`가 `하루에 하나`라는 비즈니스 규칙을 책임지는 것이 적절한가?

엘리의 리뷰에 답변을 작성하다가 개별 객체 하나가 중복(최소 2개 이상의 요소가 있어야 말이 되는)을 위한
구현까지 가지고 있어야 하나? 라는 고민에 빠졌습니다.

만약 AttendanceLog를 관리하는 컬렉션 객체(AttendanceLogs)가 중복을 책임지는 경우
조금 더 유연하게 설계될 수 있지 않을까?

출석 기록의 중복 여부를 관리하는 책임을 Log가 아닌 Logs에 주고
Log는 단순히 `출석 기록을 표현하는 객체` 역할을 수행하도록 한다면,
기존보다 확장성이 높아질 수 있다고 생각합니다.
예를 들어, 한 사람이 하루에 여러 번 출석이 가능하도록 변경되어야 한다면, Logs만 수정하면 될 것 같습니다.

### 다른 방법?

혹은, 기존 방법을 유지하면서도 로그의 성격을 더 명확하게 표현하는 방법은 어떨까요?
요구 사항에 비해 복잡도가 높아지지만, 핏한 equals가 변경에 대비하기 어렵다고 판단한다면,
아래의 구조도 고려해볼 수 있지 않을까? 라는 궁금증이 생겼습니다.

```java
abstract class Log {
    private LocalDate date;
    private LocalTime time;

    // 공통 로직
}

public class AttendanceLog extends Log {
    private Nickname nickname;

    // 출석 기록은 하루에 하나만 존재해야 하므로
    // 기존처럼 닉네임 + 날짜 기준의 equals / hashCode 유지
}

public class OutLog extends Log {
    private Nickname nickname;

    // 외출(및 복귀) 로그는 여러 개 있어도 되므로
    // 적절한 equals / hashCode 구현
}
```

이렇게 하면 각 로그가 자신만의 기준을 가질 수 있을 것 같지만, 복잡도는 많이 높아질 수 있겠네요.
이런 방식의 접근도 괜찮은지 엘리에 의견이 궁금합니다! 😄 

### 최종 결론

> Log가 책임지던 `중복을 위한 핏한 equals`를 Logs가 스스로 책임지도록 변경해보겠습니다 👍 

고민해 보다가 위 방식이 더 유연한 방식인 것 같다고 느꼈습니다.

yebink replied:
일단 잘 리팩토링 하셨습니다~💯 

말씀주신 `AttendanceLog`, `OutLog` 방식도 고려해볼만 하나, 구현 방식에 따라 로그 정보 생성 시 중복이 발생할 수도 있을 것 같네요. 지금 요구사항에서는 `AttendanceLogs` 에서 처리하는 것이 더 좋아보입니다 👍

yebink commented on src/main/java/attendance/model/AttendancesFile.java:
```diff
@@ -0,0 +1,74 @@
+    private static final int EXPECTED_COLUMN_COUNT = CSV_HEADER_FORMAT.split(COLUMN_DELIMITER).length;
+    private static final int NICKNAME_INDEX = 0;
+    private static final int ATTENDANCE_DATE_TIME_INDEX = 1;
+    private static final String DATETIME_FORMAT_PATTERN = "yyyy-MM-dd HH:mm";
```
DateTimeFormatter 는 생성 비용이 비싼 객체이기 때문에 보통 한 번 정의해두고 재사용하는 방식을 많이 사용합니다 🙂 

```suggestion
    private static final DateTimeFormatter DATETIME_FORMAT_PATTERN = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```

YehyeokBang replied:
`DateTimeFormatter`는 단순한 포맷터가 아니라, 
출력/파싱, 지역화, 연대/타임존 변환, 필드 해석 등 다양한 기능을 수행하네요.

객체 생성 시 이러한 요소들을 조합하는 과정에서 객체 검증, 복잡한 설정 처리, 불변성 유지 등의 비용이 발생하며,
동일한 포맷을 여러 번 사용할 경우 매번 생성하지 말고, 재사용하는 것이 성능 측면에서 확실히 유리하겠네요 👍 👍 

감사합니다! 반영하겠습니다!

yebink commented on src/main/java/attendance/model/AttendancesFile.java:
```diff
@@ -0,0 +1,74 @@
+        try {
+            return LocalDateTime.parse(rawDateTime, DateTimeFormatter.ofPattern(DATETIME_FORMAT_PATTERN));
+        } catch (DateTimeParseException e) {
+            throw new IllegalStateException("입력된 시간 형식이 올바르지 않습니다. 입력: %s".formatted(rawDateTime));
```
`IllegalStateException`과 `IllegalArgumentException` 의 차이는 무엇일까요? 🤔

YehyeokBang replied:
`IllegalArgumentException`
잘못된 인자(입력 값)로 인해 발생하는 예외를 의미하며, 
메서드에 전달된 값이 기대하는 범위나 형식에 맞지 않을 때 사용합니다.

`IllegalStateException`
객체의 현재 상태가 호출하려는 메서드 실행과 맞지 않을 때 발생하는 예외
즉, 메서드를 호출할 수 없는 부적절한 상태일 때 사용합니다.

YehyeokBang replied:
### 추가로 궁금한 점

`AttendancesFile.java`

파일의 내용은 프로그래머(또는 시스템 관리자)가 미리 준비해야 하는 데이터라고 생각했습니다.
프로그램이 실행되기 전에 이미 잘못된 상태일 수 있으며, 프로그램이 실행될 때 파일이 존재하지 않거나,
데이터 형식이 깨져 있다면, 이는 잘못된 `상태(State)`에서 실행되는 것이라고 판단하여
`IllegalStateException`를 사용했습니다.

이는 합리적인 선택일까요? 엘리의 의견이 궁금해요!!

yebink replied:
그러한 관점에서 보면 충분히 합리적인 선택이네요 😄

다만 해당 에러메시지만 보면 사용자로부터 입력된 값이 잘못되었다고 오해할 수 있을 것 같아요.
메시지를 수정해볼까요?

yebink commented on src/test/java/attendance/model/AttendanceBookTest.java:
```diff
@@ -0,0 +1,246 @@
+        // then
+        assertThat(attendResult.attendanceDateTime())
+                .isEqualTo(attendanceDateTime);
+        assertThat(attendResult.attendanceType() == AttendanceType.PRESENT)
```
아래와 같이 쓸 수 있습니다 😄 

```suggestion
        assertThat(attendResult.attendanceType()).isSameAs(AttendanceType.PRESENT);
```

YehyeokBang replied:
Enum을 `isEqualTo`로 검증하면 결국 `equals`를 사용하게 되는데,  
이 방식이 항상 안전한지 고민이 되어 직접 `==`로 비교하는 방식을 테스트했었습니다. 

더 좋은 방법을 알려주셔서 감사합니다! 👍 👍

yebink commented on src/test/java/attendance/model/AttendanceWarningTest.java:
```diff
@@ -0,0 +1,26 @@
+import org.junit.jupiter.api.Test;
+
+@DisplayName("출석 경고 테스트")
+class AttendanceWarningTest {
```
이 테스트는 미완성이네요! 😅

YehyeokBang replied:
놓친 부분 확인했습니다. 항상 꼼꼼한 리뷰 감사해요 😄

yebink commented on src/main/java/attendance/model/AttendanceLogs.java:
```diff
@@ -0,0 +1,126 @@
+    private void addIfOpenDay(Nickname nickname, List<AttendanceLog> completeLogs, LocalDate date) {
+        try {
+            completeLogs.add(new AttendanceLog(nickname, date));
+        } catch (IllegalArgumentException ignore) {
```
예외가 발생하면 무시하는 로직으로 보이는데요~
어떤 상황을 가정하고 작성하신 걸까요?

YehyeokBang replied:
출석할 수 있는 날이 아니라면 `AttendanceLog`를 생성할 수 없습니다. (공휴일, 주말 등)
그래서 메서드명으로 최대한 설명하려고 했습니다.
다시 보니, AttendanceLog의 책임을 모른다면, 이 의도를 파악하기 힘들다는 문제가 있는 것 같습니다.

다른 방법을 고민해보겠습니다. 👍

yebink commented:
제가 실수로 close 를 눌러서 다시 열었습니다 ㅎㅎ

### TDD와 커밋 단위에 대한 고민
의미 있는 고민이네요 👍
개인적으로 커밋 메시지에 TDD 사이클이 드러날 필요는 없다고 생각합니다. TDD는 결국 견고한 설계와 더 높은 코드 품질을 위해 선택하는 방식 중 하나일 뿐, 누군가에게 TDD 로 작성되었음 을 보여주는 행위 자체가 중요하다고 생각하지는 않습니다 😄 

벨로가 말한 2가지 규칙도 중요하다고 생각합니다! 추가로 제가 커밋 단위에 대해 중요하게 생각하는 부분이 있다면, `언제든지 롤백할 수 있는 단위로 커밋이 되었는가?` 입니다. 빌드와 테스트가 실패하지 않는 단위로 커밋을 하고, 주로 1커밋 1기능 (1리팩토링) 단위로 커밋을 하려고 합니다.

### 출력 요구 사항과 메서드의 명확성
지금 코드를 보면, 도메인에서 출력에 대한 정보를 너무 많이 알고 있네요!
코드가 단순해졌다고 하셨는데, 도메인 입장에서도 단순해졌을까요?

출력에 대한 요구사항이 추가되거나 변경되면 도메인까지 변경이 발생하게 됩니다.
만약 `EditResult` 가 여러 개 필요한 상황이 오면 어떻게 될까요?
제가 속한 팀에서는 가게의 변경이력을 관리합니다. 이때 내부 관리자용 어드민에 노출하는 이력과 사장님들에게 노출하는 이력은 서로 다릅니다.
그러면 도메인에서 `이력Result`를 여러 개 리턴하는 방식이 효율적일까요~?

yebink reviewed (changes requested):
안녕하세요 벨로 ~

전반적으로 잘 구현하셨어요 👍 
추가 피드백 몇가지 남겼으니 확인해주세요~!

YehyeokBang committed:
fix: 완료된 TODO 주석 제거

YehyeokBang committed:
refactor: 열거형 비교 테스트 방식 변경

- 기존 == 연산자로 직접 비교 후 확인하던 방식에서 isSameAs 방식으로 변경합니다.

YehyeokBang committed:
refactor: 닉네임 등록 명단 클래스명 수정

- 등록되어 있다는 의미를 포함하도록 변경합니다.
- 의미가 불분명한 메서드명을 수정합니다.
- 테스트를 위한 Fixture를 추가합니다.

YehyeokBang committed:
refactor: DateTimeFormatter 상수화

- 반복되는 출력일 경우 수집 후 출력하거나 StringBuilder를 사용하도록 수정합니다.

YehyeokBang committed:
refactor: 출석 로그 수정 DTO 의존성 제거

- 출력에 의존하지 않도록 수정합니다.
- dto 클래스임을 이름에 명시합니다.

YehyeokBang committed:
refactor: 모델의 dto 의존 제거

- 추가 및 수정 등의 명령형 기능은 명령만 수행하도록 수정합니다.
- view에 필요한 데이터는 필요한 순간에 조회하도록 수정합니다.
- 불필요해진 dto를 제거합니다.

YehyeokBang committed:
refactor: 모델 dto 의존 제거

- 모델이 출력을 위한 dto를 생성하지 않도록 수정합니다.
- dto 클래스에 dto임을 명시합니다.

YehyeokBang committed:
refactor: 입력 값 시간 파싱 메서드 위치 변경

YehyeokBang committed:
refactor: 예외를 무시하던 로직 수정

- 기존 예외를 무시하도록 구성한 로직을 직접 검증하여 추가 여부를 판단하도록 변경합니다.
- 가독성 및 안전성 향상을 위한 작업입니다.

YehyeokBang committed:
refactor: 같은 날 출석 불가 검증 로직 위치 변경

- 같은 날에 출석은 한번만 가능하다는 검증 요구사항 구현을
  개별 로그(equals 구현)가 아닌 로그의 집합이 검증하도록 수정합니다.

YehyeokBang committed:
test: Test Fixtures 적용

- 불변 및 상태를 가지지 않는 경우 재사용할 수 있도록 적용합니다.

YehyeokBang committed:
refactor: 빠진 테스트 추가 작성

- getter 또는 재정의한 equals 등을 테스트하는 코드를 추가합니다.

YehyeokBang committed:
refactor: 코드 포맷팅 수정

YehyeokBang commented:
## TDD와 커밋 단위에 대한 고민

> 언제든지 롤백할 수 있는 단위로 커밋이 되었는가? 입니다. 빌드와 테스트가 실패하지 않는 단위로 커밋을 하고, 주로 1커밋 1기능 (1리팩토링) 단위로 커밋을 하려고 합니다.

좋은 기준인 것 같습니다. 공부를 하면서 대부분 롤백할만한 상황이 없었던 것 같아서 생각하지 못했던 부분인 것 같아요.
좋은 인사이트 감사합니다 👍 

## 출력 요구 사항과 메서드의 명확성

> 지금 코드를 보면, 도메인에서 출력에 대한 정보를 너무 많이 알고 있네요!
코드가 단순해졌다고 하셨는데, 도메인 입장에서도 단순해졌을까요?
>
> 출력에 대한 요구사항이 추가되거나 변경되면 도메인까지 변경이 발생하게 됩니다.
만약 EditResult 가 여러 개 필요한 상황이 오면 어떻게 될까요?
제가 속한 팀에서는 가게의 변경이력을 관리합니다. 이때 내부 관리자용 어드민에 노출하는 이력과 사장님들에게 노출하는 이력은 서로 다릅니다.
그러면 도메인에서 이력Result를 여러 개 리턴하는 방식이 효율적일까요~?

모델이 애초에 출력과 관련된 것을 가지지 않는 방향이 효율적일 것 같아요.
출력 요구 사항이 변경될 때마다 중요한 모델에 수정이 발생하고,
하나를 만들면, 그 이후로도 쉽게 만들어질 것 같네요.
또한, 노출되면 안될 것들을 모델에서 직접 모두 관리하게 되는 것은 부담될 것 같아요.

그래서 모델에서 dto를 만들던 방법에서 dto를 없애거나, dto를 Controller에서 만드는 방식으로 수정했습니다.

## 피드백 반영

- [x]  명확한 클래스명 사용
    - dto인 경우 dto임을 명시
    - 크루 명단 클래스명을 등록해둔다는 의미를 표현
- [x] 모델 dto 의존 제거
    - 추가하거나 수정하는 메서드는 그 작업만 수행하도록 변경하고, 모델의 dto 반환 로직을 제거합니다.
- [x]  equals / hashCode 변경
    - 각 로그 객체가 중복을 책임지던 방식에서 로그의 집합이 책임지도록 변경합니다.
    - [refactor: 같은 날 출석 불가 검증 로직 위치 변경](https://github.com/woowacourse/java-attendance/pull/116/commits/3c57a61e22603b06454aebfccbd367115ae7c659)
- [x]  DateTimeFormatter 재사용
- [x]  Enum 테스트 방식 변경 `== + isTrue()` → `isSameAs()`
  - [refactor: 열거형 비교 테스트 방식 변경](https://github.com/woowacourse/java-attendance/pull/116/commits/7e85936dd49498f62384192b18003986e5c42fe6)
- [x]  예외 발생 무시 로직을 더욱 명확하게 변경
  - [refactor: 예외를 무시하던 로직 수정](https://github.com/woowacourse/java-attendance/pull/116/commits/144399ffc4dd7192c561ce94a89b1f2ef721f30c)
- [x]  완료된 TODO 제거 및 테스트 코드 작성
- [x] 불변 객체 위주로 TestFixtures 도입

이번 리뷰도 잘 부탁드립니다. 감사합니다. 👍 👍

yebink commented on src/test/java/attendance/model/TestFixtures.java:
```diff
@@ -0,0 +1,26 @@
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+
+public class TestFixtures {
```
👍

yebink commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -1,258 +1,224 @@
-                determineAttendanceType(beforeAttendance),
-                determineAttendanceType(modifidedAttendance)
-        );
+    private List<AttendanceLogDto> mapToAttendanceLogDtos(AttendanceBook attendanceBook,
```
domain을 dto로 매핑하는 행위는 dto 내부에서 가지는 것을 추천드립니다!

dto의 목적에는 controller로부터 domain을 보호하는 데이터 캡슐화 기능도 있습니다.
dto를 구성하는 로직을 controller에 담는다면, `AttendanceBook`이나 `Nickname` 같은 도메인들이 controller에 모두 노출되게 됩니다.
controller에서 model 을 몰라야 한다는 것은 아니지만, dto를 사용하신다면 목적에 맞게 사용하셨으면 좋겠습니다~ 😄

YehyeokBang replied:
> DTO의 목적 중 하나는 도메인 모델을 보호하고 데이터를 캡슐화하는 것이다.

말씀해주신 것처럼 DTO 내부에서 정적 팩토리 메서드를 활용하여 변환 책임을 DTO가 가지도록 하면,
컨트롤러는 흐름만을 제어하고 변환 로직은 DTO가 관리할 수 있어 더 깔끔한 구조가 되는 것 같습니다.

DTO로 필요한 값만 추출하여 도메인을 보호할 수 있고,
뷰의 변경이 발생하더라도 영향이 DTO까지만 전파되므로,
도메인 모델(또는 컨트롤러)이 불필요하게 영향을 받는 것을 방지할 수 있겠네요!

DTO의 본래 역할을 강화하면서 유지보수성을 높이는 좋은 방향인 것 같습니다. 반영하겠습니다! 👍 👍

yebink commented on src/main/java/attendance/dto/EditResultDto.java:
```diff
@@ -0,0 +1,14 @@
+import java.time.LocalDate;
+import java.time.LocalTime;
+
+public record EditResultDto(
```
미사용 dto는 제거해주세요!

yebink reviewed (changes requested):
안녕하세요 벨로~

피드백 잘 반영해주셨네요! 💯 
몇가지 질문에 대한 답변과 추가 피드백 드리니 확인해주세요 😄

YehyeokBang committed:
fix: 미사용 dto 제거

YehyeokBang committed:
refactor: 출석 로그 파일 관련 예외 메시지 수정

- 사용자의 실수가 아닌 시스템의 문제라는 것을 인지할 수 있도록 수정합니다.

YehyeokBang committed:
refactor: dto 매핑을 스스로 할 수 있도록 수정

- 기존 컨트롤러에서 매핑하던 방식을 dto 내부에서 진행하도록 수정합니다.

YehyeokBang commented:
## 리뷰 반영

말씀해주신 DTO 관련 피드백이 더 나은 방법이라고 생각해서 반영했습니다!

항상 꼼꼼한 리뷰 덕분에 근거 확립에 많은 도움이 되었습니다. 감사합니다. 😄

yebink reviewed (approved):
안녕하세요 벨로~

출석 미션 고생 많으셨습니다!
추가로 피드백 드릴 부분이 없어 머지하겠습니다 😄
